### ISSUE TITLE
Coachmark doesn't follow target element

### ISSUE THREAD

**Issue by dwootton** [microsoft/fluentui#18149]
### Environment Information

I've noticed it in "office-ui-fabric-react": "^7.166.1"; however, it's present on the docs site so it's likely present in multiple versions.

#### Actual behavior:

When scrolling (or if an element is displaced from initial load due to page layout changes) the coachmark is stuck at the original placement of the element rather than updating it's location to the new element placement. 
![Weird Offset](https://user-images.githubusercontent.com/30129761/117858172-75456580-b242-11eb-902c-ab4318022275.gif)

#### Expected behavior:

I would expect the coachmark to follow the element that it is attached to. The current jumping makes the user confused as to what they should be clicking and as it applies to both the mark itself and the info box, it makes the component difficult to use.

### Priorities and help requested:

Are you willing to submit a PR to fix? I'm not too familiar with the code base, but if there's an easy fix I'd be willing to help

Requested priority: Blocking

Products/sites affected: Internal Analytics site for MS Edge

**Comment by gouttierre** [comment]
@dwootton - Thanks for filing this issue and I see you are willing to help submit a PR for the fix (if confirmed) -  providing we give  guidance - That is great to hear! 

@miroslavstastny - Would you be able to confirm if this is a regression, or if this behavior is an issue?
cc @smhigley

**Comment by miroslavstastny** [comment]
Even though the original report says @dwootton noticed the issue in v7, I am not able to reproduce it there, but can reproduce in 8.17.4.
-> Regression in v8

**Comment by miroslavstastny** [comment]
The bug is in `PositioningContainer.tsx` - `dismissOnScroll` is called but it does not reposition as `positions` is undefined.

The reason is the event listener for scroll is added only on mount with `dismissOnScroll` having `positions` in its closure:
https://github.com/microsoft/fluentui/blob/47b42efbbe9e869864973f3b2cb717e6112b8106/packages/react/src/components/Coachmark/PositioningContainer/PositioningContainer.tsx#L240-L241

One solution would be to fix the dependencies in the `useEffect` (adding `dismissOnScroll` as a dependency fixes the issue, but the callback is unstable -> event handler is removed/added on each render). Other solution would be to store `positions` to a ref and use the ref in `dismissOnScroll`.
